extensions/guacamole-vault/modules/guacamole-vault-ksm/src/main/java/org/apache/guacamole/vault/ksm/user/KsmDirectoryService.java

```diff
@@ -0,0 +1,241 @@
+            // so validate it immediately
+            if(isBase64(ksmConfigValue)) {
+
+                // Attempt to validate the config as a base64-econded KSM config blob
+                try {
+                    KsmConfig.parseKsmConfig(ksmConfigValue);
+
+                    // If it validates, the entity can be returned as-is
+                    return;
+                }
+
+                catch (GuacamoleException exception) {
+
+                    // If the parsing attempt fails, throw a translatable error for display
+                    // on the frontend
+                    throw new TranslatableGuacamoleClientException(
+                            "Invalid base64-encoded JSON KSM config provided for "
+                            + KsmAttributeService.KSM_CONFIGURATION_ATTRIBUTE + " attribute",
+                            "CONNECTION_GROUP_ATTRIBUTES.ERROR_INVALID_KSM_CONFIG_BLOB");
+                }
+            }
+
+            else {
+
+                try {
+
+                    // Create an initially empty storage to be populated using the one-time token
+                    InMemoryStorage storage = new InMemoryStorage();
+
+                    // Populate the in-memory storage using the one-time-token
+                    SecretsManager.initializeStorage(storage, ksmConfigValue, null);
+
+                    // Create an options object using the values we extracted from the one-time token
+                    SecretsManagerOptions options = new SecretsManagerOptions(
+                        storage, null,
+                        configurationService.getAllowUnverifiedCertificate());
+
+                    // Attempt to fetch secrets using the options we created. This will both validate
+                    // that the configuration works, and potentially populate missing fields that the
+                    // initializeStorage() call did not set.
+                    SecretsManager.getSecrets(options);
+
+                    // Create a map to store the extracted values from the KSM storage
+                    Map<String, String> configMap = new HashMap<>();
+
+                    // Go through all the expected fields, extract from the KSM storage,
+                    // and write to the newly created map
+                    EXPECTED_KSM_FIELDS.forEach(configKey -> {
+
+                        // Only write the value into the new map if non-null
+                        String value = storage.getString(configKey);
+                        if (value != null)
+                            configMap.put(configKey, value);
+
+                    });
+
+                    // JSON-encode the value, and then base64 encode that to get the format
+                    // that KSM would expect
+                    String jsonString = new ObjectMapper().writeValueAsString(configMap);
+                    String base64EncodedJson = Base64.getEncoder().encodeToString(jsonString.getBytes());
```

Comment:
This will use the platform's default charset, whatever that happens to be. We should instead explicitly specify UTF-8 (assuming that's the intent).